src/device_handle/async_api.rs

```diff
@@ -0,0 +1,266 @@
+    // `poll()` is called.
+    extern "system" fn transfer_cb(transfer: *mut ffi::libusb_transfer) {
+        // Safety: libusb should never make this null, so this is fine
+        let transfer = unsafe { &mut *transfer };
+
+        // sanity
+        debug_assert_eq!(
+            transfer.transfer_type,
+            ffi::constants::LIBUSB_TRANSFER_TYPE_BULK
+        );
+
+        let transfer: *mut AsyncTransfer<C> = transfer.user_data.cast();
+        let transfer = unsafe { &mut *transfer };
+        // Mark transfer as completed
+        transfer
+            .completed_queue
+            .lock()
+            .unwrap()
+            .push_back(transfer.pool_id);
+    }
+
+    // Step 3 of async API
+    fn submit(self: &mut Transfer<C>) -> Result<(), AsyncError> {
+        let transfer_struct = self.ptr;
+        let errno = unsafe { ffi::libusb_submit_transfer(transfer_struct.as_ptr()) };
+
+        use ffi::constants::*;
+        use AsyncError as E;
+        match errno {
+            0 => Ok(()),
+            LIBUSB_ERROR_NO_DEVICE => Err(E::Disconnected),
+            LIBUSB_ERROR_BUSY => {
+                unreachable!("We shouldn't be calling submit on transfers already submitted!")
+            }
+            LIBUSB_ERROR_NOT_SUPPORTED => Err(E::Other("Transfer not supported")),
+            LIBUSB_ERROR_INVALID_PARAM => Err(E::Other("Transfer size bigger than OS supports")),
+            _ => Err(E::Errno("Error while submitting transfer: ", errno)),
+        }
+    }
+}
+impl<C: UsbContext> Drop for AsyncTransfer<C> {
+    fn drop(&mut self) {
+        // TODO: Figure out how to destroy transfers, which is step 5 of async API.
+        todo!()
+    }
+}
+
+/// Represents a pool of asynchronous transfers, that can be polled to completion
+pub struct AsyncPool<C: UsbContext> {
+    /// Contains the pool of AsyncTransfers
+    pool: Vec<Transfer<C>>,
+    /// Contains the idxs of transfers in `pool` that have completed
+    completed: CompletedQueue,
+    device: Arc<DeviceHandle<C>>, // TODO: We hold refs to this, do we need it to be pinned?
+}
+impl<C: UsbContext> AsyncPool<C> {
+    pub fn new_bulk(
+        device: DeviceHandle<C>,
+        endpoint: u8,
+        read_timeout: Duration,
+        buffers: impl IntoIterator<Item = Vec<u8>>,
+    ) -> Result<Self, AsyncError> {
+        let buffers = buffers.into_iter();
+        let mut pool = Vec::with_capacity(buffers.size_hint().0);
+        let completed = Arc::new(Mutex::new(VecDeque::with_capacity(buffers.size_hint().0)));
+        let device = Arc::new(device);
+
+        for (id, buf) in buffers.into_iter().enumerate() {
+            let mut transfer: Transfer<C> = AsyncTransfer::new_bulk(
+                id,
+                completed.clone(),
+                device.clone(),
+                endpoint,
+                buf,
+                read_timeout,
+            );
+            transfer.submit()?;
```

Comment:
Is there any way to avoid this? To do otherwise would be to have either a `first_time_polling` flag or a list of transfers that must initially be submitted. 

I haven't thought at all about a write api - I have never used libusb's async api to do writes (or anything other than bulk reads)

Reply:
This is why I think submitting transfers should be separate than initialization, especially for writes. For reads I think it also may be a good idea to make submitting transfers explicit, for example you may want to stop submitting new transfers but let the existing ones complete if you know ahead of time how much data you're trying to read.